infra/base-images/base-builder/jcc/jcc.go

```diff
@@ -174,6 +177,55 @@ func compile(bin string, args []string) (int, string, string) {
 	return cmd.ProcessState.ExitCode(), outb.String(), errb.String()
 }
 
+func FindTargetFile(args []string) string {
+	// Finds the fuzz target file by file extension.
+	suffixes := []string{".cpp", ".cc", ".cxx", ".c++", ".c"}
+	for _, arg := range args {
+		if slices.Contains(suffixes, strings.ToLower(filepath.Ext(arg))) {
+			return filepath.Base(arg)
+		}
+	}
+	// Uses a time stamp as AST filename if no target file is found.
+	return time.Now().Format("20060102_150405.000")
+}
+
+func RemoveIfEmpty(filepath string) {
+	// Removes filepath if it is empty.
+	info, _ := os.Stat(filepath)
+	if info.Size() == 0 { os.Remove(filepath) }
+}
+
+func GenerateAST(bin string, args []string) (int, string, string) {
+	// Generates AST.
+	newArgs := append(args, "-Xclang", "-ast-dump=json", "-fsyntax-only")
+
+	targetFile := FindTargetFile(args)
+	filePath := filepath.Join("/out", fmt.Sprintf("%s.ast", targetFile))
+	outFile, err := os.Create(filePath)
+	if err != nil {
+		fmt.Println(err)
+	}
+	defer outFile.Close()
+
+	cmd := exec.Command(bin, newArgs...)
+	var errb bytes.Buffer
+	cmd.Stdout = outFile
+	cmd.Stderr = &errb
+	cmd.Run()
+	// Cleans up useless empty AST files. They are generated when the cmd
+	// does not contain C/CPP file.
+	RemoveIfEmpty(filePath)
+	return cmd.ProcessState.ExitCode(), "", errb.String()
+}
+
+func compile(bin string, args []string) (int, string, string) {
+	// Generate AST.
+	retCode, _, stderr := GenerateAST(bin, args)
```

Comment:
Can we also control this with an env var? i.e. off by default, but if we need the AST, we set an the env var.

Reply:
Sure, I can implement that.
I will only set the env var when a C/CPP file appears in the command.

Reply:
Added ENV var `JCC_GENERATE_AST` and only generate ASTs when it is `"1"`.